src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -835,28 +1025,57 @@ SDCPUThread::populateDependencies(shared_ptr<InflightInst> inst_ptr)
 }
 
 void
-SDCPUThread::populateUses(shared_ptr<InflightInst> inst_ptr) {
+SDCPUThread::populateUses(shared_ptr<InflightInst> inst_ptr)
+{
     const StaticInstPtr static_inst = inst_ptr->staticInst();
     const int8_t num_dsts = static_inst->numDestRegs();
     for (int8_t dst_idx = 0; dst_idx < num_dsts; ++dst_idx) {
         const RegId& dst_reg = static_inst->destRegIdx(dst_idx);
 
-        InflightInst::DataSource source = {inst_ptr, dst_idx};
-        lastUses[dst_reg] = source;
+        lastUses[dst_reg] = {inst_ptr, dst_idx};
         DPRINTF(SDCPUDeps, "%d is producer of reg[%d]\n", inst_ptr->seqNum(),
                 dst_reg.index());
     }
 }
 
-bool
-SDCPUThread::hasNextPC()
+void
+SDCPUThread::predictCtrlInst(shared_ptr<InflightInst> inst_ptr)
 {
-    // TODO check if we're still running
+    DPRINTF(SDCPUBranchPred, "Requesting branch predictor for control\n");
 
-    if (inflightInsts.empty()) return true;
+    const InstSeqNum seqnum = inst_ptr->seqNum();
+    inst_ptr->addSquashCallback([this, seqnum] {
+        _cpuPtr->getBranchPredictor()->squash(seqnum, threadId());
+    });
 
-    const shared_ptr<InflightInst> inst_ptr = inflightInsts.back();
-    return !inst_ptr->staticInst()->isControl() || inst_ptr->isComplete();
+    const weak_ptr<InflightInst> weak_inst = inst_ptr;
+    auto callback = [this, weak_inst] (BPredUnit* pred) {
+        onBranchPredictorAccessed(weak_inst, pred);
+    };
+
+    _cpuPtr->requestBranchPredictor(callback);
+}
+
+void
+SDCPUThread::squashUpTo(shared_ptr<InflightInst> inst_ptr, bool rebuild_map)
+{
+    size_t count = 0;
+    while (inflightInsts.back() != inst_ptr) {
+        inflightInsts.back()->notifySquashed();
+        inflightInsts.pop_back();
+        ++count;
+    }
+
+    DPRINTF(SDCPUThreadEvent, "%d instructions squashed.\n", count);
+
+    if (rebuild_map) {
+        for (shared_ptr<InflightInst>& it : inflightInsts) {
+            populateUses(it);
```

Comment:
Well there's two ways of going at it: Iterate backwards on inflightInsts to find the last writer of all registers, or go forwards, and replace as we go. Both involve iteration, and I don't think it's trivially avoidable, since the map is invalidated when we squash things without clearing the entire inflightInsts buffer, and does not keep any history of what came before.

Reply:
Couldn't we just clear the entire thing? If we squash *all* instructions, then only the instructions that have been committed are valid.

Reply:
Well we only squash the ones that came after the branch that was mispredicted...

Reply:
The other thing is that squashes should not be a super common event anyway.

Reply:
MCCF right?

Reply:
your branch misprediction penalty is that your universe will simulate more slowly lol

Reply:
Anything else before we merge this?

Reply:
No, I don't think so.